src/scripts/release.zig

```diff
@@ -530,106 +534,286 @@ fn build_tigerbeetle(shell: *Shell, info: VersionInfo, dist_dir: std.fs.Dir) !vo
-    defer body_file.close();
+    var releases = multiversioning.ListU32{};
+    var checksums = multiversioning.ListU128{};
+    var offsets = multiversioning.ListU32{};
+    var sizes = multiversioning.ListU32{};
+    var flags_ = multiversioning.ListFlag{};
+    var git_commits = multiversioning.ListGitCommit{};
+    var release_client_mins = multiversioning.ListU32{};
+
+    // Extract the old current release - this is the release that was the current release, and not
+    // embedded in the past pack.
+    const old_current_release = header.current_release;
+    const old_current_release_output_name = try std.fmt.allocPrint(
+        shell.arena.allocator(),
+        "{s}-{}",
+        .{ tigerbeetle_name_binary, old_current_release },
+    );
 
-    try body_file.writeAll(past_binary_contents);
+    if (macos) {
+        const cpu_type = if (std.mem.eql(u8, target, "aarch64-macos"))
+            std.macho.CPU_TYPE_ARM64
+        else
+            std.macho.CPU_TYPE_X86_64;
+
+        const cpu_subtype = if (cpu_type == std.macho.CPU_TYPE_ARM64)
+            std.macho.CPU_SUBTYPE_ARM_ALL
+        else
+            std.macho.CPU_SUBTYPE_X86_64_ALL;
+
+        try macos_universal_binary_extract(
+            shell,
+            tigerbeetle_name_zip,
+            .{ .cpu_type = cpu_type, .cpu_subtype = cpu_subtype },
+            old_current_release_output_name,
+        );
+    } else {
+        try shell.exec("{llvm_objcopy} --enable-deterministic-archives --keep-undefined" ++
+            " --remove-section .tb_mvb --remove-section .tb_mvh " ++
+            "{tigerbeetle_name_zip} {old_current_release_output_name}", .{
+            .llvm_objcopy = llvm_objcopy,
+            .tigerbeetle_name_zip = tigerbeetle_name_zip,
+            .old_current_release_output_name = old_current_release_output_name,
+        });
+    }
 
-    const git_commit: [20]u8 = blk: {
-        var commit_bytes: [20]u8 = std.mem.zeroes([20]u8);
-        for (0..@divExact(multiversion_epoch_commit.len, 2)) |i| {
-            const byte = try std.fmt.parseInt(u8, multiversion_epoch_commit[i * 2 ..][0..2], 16);
-            commit_bytes[i] = byte;
+    // It's important to verify the previous current_release checksum - it can't be verified at
+    // runtime by multiversioning.zig, since it relies on objcopy to extract.
+    assert(header.current_checksum == try checksum_file(
+        shell,
+        old_current_release_output_name,
+        multiversion_binary_size_max,
+    ));
+
+    const old_current_release_size: u32 = @intCast((try (try shell.cwd.openFile(
+        old_current_release_output_name,
+        .{ .mode = .read_only },
+    )).stat()).size);
+
+    // FIXME: This can vary by platform; so we could create a binary for Linux that has 10 past
+    // versions and one for Windows that only has 8. What do we want to do here?
+    const past_count = blk: {
+        const max_size = multiversion_binary_platform_size_max(.{
```

Comment:
Hm, this seems wrong, shouldn't we also count the _current_ size towards the limit?

Reply:
Yes - good catch!